jwst/lib/pipe_utils.py

```diff
@@ -76,3 +83,65 @@ def is_irs2(model):
         return True
     else:
         return False
+
+
+def match_nans_and_flags(input_model):
+    """Ensure data, error, variance, and DQ are marked consistently for invalid data.
+
+    Invalid data is assumed to be any pixel set to NaN in any one of the
+    data, error, or variance arrays, or else set to the DO_NOT_USE flag
+    in the DQ array.
+
+    The input model is updated in place with NaNs or DO_NOT_USE flags, as
+    appropriate, at all invalid data locations.
+
+    Parameters
+    ----------
+    input_model : DataModel
+        Input model containing some combination of data, dq, err, var_rnoise,
+        var_poisson, and var_flat extensions. These extensions must all have
+        matching dimensions if present.
+    """
+    # Build up the invalid data flags from each available data extension.
+    is_invalid = None
```

Comment:
Given the silent issue with the tso data for outlier detection. Would you add a `isinstance` check here to make sure the input is a `DataModel`?

Reply:
Sure.  I'll make that condition throw an error, since that's likely a code issue.

Reply:
Done, but I also had to allow datamodel ObjectNode instances as well - slits in MultSlitModels do not retain the SlitModel data type when they are appended to the `slits` list node.

Reply:
Thanks!